p11-kit/proxy.c

```diff
@@ -279,29 +277,59 @@ proxy_list_slots (Proxy *py, Mapping *mappings, unsigned int n_mappings)
 
 		if (count > 0) {
 			Mapping *new_mappings;
+			bool *is_new_mapping;
+			CK_SLOT_ID max_id = 0;
 
+			is_new_mapping = calloc (count, sizeof(bool));
+			return_val_if_fail (is_new_mapping != NULL, CKR_HOST_MEMORY);
 			new_mappings = realloc (py->mappings, sizeof (Mapping) * (py->n_mappings + count));
 			return_val_if_fail (new_mappings != NULL, CKR_HOST_MEMORY);
 			py->mappings = new_mappings;
 
-			/* And now add a mapping for each of those slots */
+			/* Reuse the existing mapping if any */
 			for (i = 0; i < count; ++i) {
-				/* Reuse the existing mapping if any */
 				for (j = 0; j < n_mappings; ++j) {
 					/* cppcheck-suppress nullPointer symbolName=mappings */
 					/* false-positive: https://trac.cppcheck.net/ticket/9573 */
 					if (mappings[j].funcs == funcs &&
-					    mappings[j].real_slot == slots[i])
+					    mappings[j].real_slot == slots[i]) {
+						py->mappings[py->n_mappings].funcs = funcs;
+						py->mappings[py->n_mappings].real_slot = slots[i];
+						py->mappings[py->n_mappings].wrap_slot =
+							mappings[j].wrap_slot;
+						++py->n_mappings;
 						break;
+					}
 				}
-				py->mappings[py->n_mappings].funcs = funcs;
-				py->mappings[py->n_mappings].wrap_slot =
-					(n_mappings == 0 || j == n_mappings) ?
-					py->n_mappings + MAPPING_OFFSET :
-					mappings[j].wrap_slot;
-				py->mappings[py->n_mappings].real_slot = slots[i];
-				++py->n_mappings;
+				if (n_mappings == 0 || j == n_mappings)
+					is_new_mapping[i] = true;
```

Comment:
can we simply maintain max_id in the Proxy structure, and increment and assign a new ID here?

Reply:
It would be better and make things easier. I tried to make it to work without changing the structures.

I'll change this.

Reply:
I moved the ``max_id`` to the proxy structure as ``last_id``, which holds the last assigned slot ID. Then we can assign unique ID's and store the used value in ``proxy->last_id``